src/presetsController.ts

```diff
@@ -0,0 +1,844 @@
+    interface PresetItem extends vscode.QuickPickItem {
+      preset: string;
+    }
+    const items: PresetItem[] = presets.filter(_preset => !_preset.hidden).map(
+        _preset => ({
+          label: _preset.displayName || _preset.name,
+          description: _preset.description,
+          preset: _preset.name
+        }),
+    );
+    items.push({
+      label: localize('add.new.preset', 'Add a New Preset...'),
+      preset: PresetsController._addPreset
+    });
+    const chosenPresets = await vscode.window.showQuickPick(items, options);
+    if (util.isArray<PresetItem>(chosenPresets)) {
+      return chosenPresets.map(_preset => _preset.preset);
+    }
+    return chosenPresets?.preset;
+  }
+
+  async selectConfigurePreset(): Promise<boolean> {
+    preset.expandVendorForConfigurePresets();
+
+    const presets = preset.configurePresets().concat(preset.userConfigurePresets()).filter(
+      _preset => {
+        const supportedHost =  (_preset.vendor as preset.Vendor_VsSettings)?.['microsoft.com/VisualStudioSettings/CMake/1.0']?.hostOS;
+        const osName = this.getOsName();
+        if (supportedHost) {
+          if (util.isString(supportedHost)) {
+            return supportedHost === osName;
+          } else {
+            return supportedHost.includes(osName);
+          }
+        } else {
+          return true;
+        }
+      }
+    );
+
+    log.debug(localize('start.selection.of.config.presets', 'Start selection of configure presets. Found {0} presets.', presets.length));
+
+    log.debug(localize('opening.config.preset.selection', 'Opening configure preset selection QuickPick'));
+    const placeHolder = localize('select.active.config.preset.placeholder', 'Select a configure preset for {0}', this.folder.name);
+    const chosenPreset = await this.showPresetSelector(presets, { placeHolder });
+    if (!chosenPreset) {
+      log.debug(localize('user.cancelled.config.preset.selection', 'User cancelled configure preset selection'));
+      return false;
+    } else if (chosenPreset === this._cmakeTools.configurePreset?.name) {
+      return true;
+    } else if (chosenPreset === '__addPreset__') {
+      await this.addConfigurePreset();
+      return false;
+    } else {
+      log.debug(localize('user.selected.config.preset', 'User selected configure preset {0}', JSON.stringify(chosenPreset)));
+      await this.setConfigurePreset(chosenPreset);
+      return true;
+    }
+  }
+
+  async setConfigurePreset(presetName: string): Promise<void> {
+    // Load the configure preset into the backend
+    await vscode.window.withProgress(
+        {
+          location: vscode.ProgressLocation.Notification,
+          title: localize('loading.config.preset', 'Loading configure preset {0}', presetName),
+        },
+        () => this._cmakeTools.setConfigurePreset(presetName),
+    );
+    await vscode.window.withProgress(
```

Comment:
When I change configure presets but have a build preset from another configure preset active, it looks like the status bar does not update and still shows the build preset from the old configure preset as active.  Should we clear it?

Reply:
Yeah I only added an error message in that case. I think we should clear it.